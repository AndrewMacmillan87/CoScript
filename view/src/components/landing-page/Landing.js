import React from "react";
import {
    BrowserRouter as Router,
    Route,
    withRouter,
    Link,
} from "react-router-dom";

import NavBar from "../NavBar";
import Contact from "./Contact";

// Redux
import { connect } from "react-redux";
import { sendMessage, loadUserProfile } from "../../redux/actions/profile";

import PropTypes from "prop-types";

import "./landing.css";

export class Landing extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            bannerID: 0,
            msgText: "",
            submitted: false,
        };
        this.showSubmitSpinner = this.showSubmitSpinner.bind(this);
        this.removeSubmitSpinner = this.removeSubmitSpinner.bind(this);
        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    /**
     * Sets the browser tab title and scrolls to
     * the top of the page when the component mounts
     */
    componentDidMount() {
        document.title = "CoScript - Project Page";
        window.scrollTo(0, 0);
    }

    /**
     * Sets the 'submitted' state property to
     * true so a loading spinner can be displayed
     */
    showSubmitSpinner() {
        this.setState({
            submitted: true,
        });
    }

    /**
     * Sets the 'submitted' state property to
     * false so a loading spinner can be removed
     */
    removeSubmitSpinner() {
        this.setState({
            submitted: false,
        });
    }

    /**
     * Sets the state property 'msgText' to
     * the value of the event target
     *
     * @param {Object} e - event object
     */
    onChange(e) {
        this.setState({
            msgText: e.target.value,
        });
    }

    /**
     * Sends a message to the project creators personal
     * profile message inbox. Redirects the user to
     * login if there are not logged in
     *
     * @param {Object} e - event object
     */
    async onSubmit(e) {
        e.preventDefault();
        this.showSubmitSpinner();
        const { msgText } = this.state;
        const { user, projectData, sendMessage, history } = this.props;
        if (!user) {
            history.push("/login");
        } else {
            const success = await sendMessage(
                msgText,
                user.userId,
                projectData.title,
                projectData.projectId,
                projectData.user
            );
            if (success) {
                this.removeSubmitSpinner();
            } else {
                setTimeout(() => {
                    this.removeSubmitSpinner();
                }, 1000);
            }
        }
    }

    render() {
        const { submitted } = this.state;
        const { projectData, loadUserProfile, history } = this.props;

        const sky = require("../../img/sky-example.jpg");
        const spinner = require("../../img/loading-black.svg");

        return (
            <main className='project-landing-page'>
                <NavBar />
                <header className='project-banner-container'>
                    <div className='banner-container'>
                        <div className='banner-details'>
                            <h1 className='project-banner-heading'>
                                {projectData ? projectData.title : ""}
                            </h1>
                            <div className='project-gravatar-container'>
                                <div className='project-gravatar'>
                                    <img
                                        className='project-gravatar-img'
                                        src={
                                            projectData
                                                ? projectData.userData.avatar
                                                : spinner
                                        }
                                        alt="User's Gravatar"
                                    />
                                    <button
                                        className='users-name'
                                        onClick={async () => {
                                            if (projectData) {
                                                await loadUserProfile(
                                                    projectData.user
                                                );
                                                const {
                                                    userProfile,
                                                } = this.props;
                                                history.push(
                                                    `/profile/user/${userProfile.id}`
                                                );
                                            }
                                        }}>
                                        {projectData &&
                                            projectData.userData.name}
                                    </button>
                                </div>
                            </div>
                        </div>
                    </div>
                    <img
                        className='project-cover-img'
                        src={sky}
                        alt='Project cover'
                    />
                </header>
                <Router>
                    <section className='project-information-container'>
                        <div className='project-tab-bar'>
                            <Link
                                className='landing-page-tab'
                                to={
                                    projectData
                                        ? `/landing${projectData.url}`
                                        : ""
                                }>
                                Description
                            </Link>
                            <p className='tab-divider'>|</p>
                            <Link
                                className='landing-page-tab'
                                to={
                                    projectData
                                        ? `/landing${projectData.url}/contact`
                                        : ""
                                }>
                                Contact
                            </Link>
                        </div>
                        <Route
                            exact
                            path={projectData && `/landing${projectData.url}`}
                            render={() => (
                                <div className='project-description'>
                                    {projectData
                                        ? projectData.description
                                        : "A description has not been provided for this project"}
                                </div>
                            )}
                        />
                        <Route
                            exact
                            path={
                                projectData &&
                                `/landing${projectData.url}/contact`
                            }
                            render={() => (
                                <Contact
                                    onChange={this.onChange}
                                    onSubmit={this.onSubmit}
                                    submitted={submitted}
                                />
                            )}
                        />
                    </section>
                </Router>
            </main>
        );
    }
}

Landing.propTypes = {
    user: PropTypes.object,
    userProfile: PropTypes.object,
    projectData: PropTypes.object,
    projectAvatar: PropTypes.object,
    sendMessage: PropTypes.func.isRequired,
    loadUserProfile: PropTypes.func.isRequired,
};

const mapProps = (state) => ({
    user: state.auth.user,
    userProfile: state.auth.userProfile,
    projectData: state.documents.projectData,
});

export default connect(mapProps, {
    sendMessage,
    loadUserProfile,
})(withRouter(Landing));
