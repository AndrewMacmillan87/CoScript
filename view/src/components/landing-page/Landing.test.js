import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import { Landing } from "./Landing";

const projectData = require("../../redux/tests/actions/data/documents/project.json");
const userData = require("../../redux/tests/actions/data/auth/user.json");

describe("Landing component", () => {
    let props, mockLoadUserProfile;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockLoadUserProfile = jest.fn();
        props = {
            user: userData,
            userProfile: userData.profile,
            projectData: projectData,
            projectAvatar: {},
            sendMessage: jest.fn(),
            loadUserProfile: mockLoadUserProfile,
            history: { push: jest.fn() },
        };
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Landing {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Landing {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a main element with a className of project-landing-page", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Landing {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("main").length).toEqual(1);
        expect(
            wrapper.find({ className: "project-landing-page" }).length
        ).toEqual(1);
    });

    it("should render a button elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Landing {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("button").length).toEqual(1);
    });

    it("should render 2 img elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Landing {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("img").length).toEqual(2);
    });

    it("should render a section element with a className of project-information-container", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Landing {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(1);
        expect(
            wrapper.find({ className: "project-information-container" }).length
        ).toEqual(1);
    });

    it("should call `loadUsersProfile()` when the profile link is clicked", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Landing {...props} />
                </Router>
            </Provider>
        );
        wrapper.find("button").at(0).simulate("click");
        await expect(mockLoadUserProfile).toBeCalled();
    });
});
