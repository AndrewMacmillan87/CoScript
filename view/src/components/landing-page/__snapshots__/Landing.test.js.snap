// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing component renders correctly 1`] = `
<main
  className="project-landing-page"
>
  <nav
    className="navbar"
  />
  <header
    className="project-banner-container"
  >
    <div
      className="banner-container"
    >
      <div
        className="banner-details"
      >
        <h1
          className="project-banner-heading"
        >
          Test Title
        </h1>
        <div
          className="project-gravatar-container"
        >
          <div
            className="project-gravatar"
          >
            <img
              alt="User's Gravatar"
              className="project-gravatar-img"
              src="http://www.gravatar.com/avatar/123456789"
            />
            <button
              className="users-name"
              onClick={[Function]}
            >
              Test Name
            </button>
          </div>
        </div>
      </div>
    </div>
    <img
      alt="Project cover"
      className="project-cover-img"
      src="sky-example.jpg"
    />
  </header>
  <section
    className="project-information-container"
  >
    <div
      className="project-tab-bar"
    >
      <a
        className="landing-page-tab"
        href="/landing/projects/5e58667d902d38559c802b14"
        onClick={[Function]}
      >
        Description
      </a>
      <p
        className="tab-divider"
      >
        |
      </p>
      <a
        className="landing-page-tab"
        href="/landing/projects/5e58667d902d38559c802b14/contact"
        onClick={[Function]}
      >
        Contact
      </a>
    </div>
  </section>
</main>
`;
